commerce/bookstore/python/load_data.py

```diff
@@ -0,0 +1,152 @@
+            for loader in loaders.loaders_list:  # Iterating through the list of classes to import all data
+                if debug: print("Loading from [" + loader("").file + "] into TypeDB ...")
+                res.append(load_data_into_typedb(loader, session))  # Call loader to load data with session
+            print("\nData loading complete!")
+            if debug: print("We have inserted the following nu,ber of items:", res)
+    return res
+
+
+def has_existing_data():  # Checking whether the DB already has the schema and the data loaded
+    with TypeDB.core_client(config.typedb_server_addr) as client:  # Establishing connection
+        with client.session(config.db, SessionType.SCHEMA) as session:  # Access data in the database
+            with session.transaction(TransactionType.READ) as transaction:  # Open transaction to read
+                try:
+                    typeql_read_query = "match $b isa book, has ISBN $x; get $x; limit 3;"
+                    transaction.query().match(typeql_read_query)
+                    print("The DB contains the schema and loaded data already.")
+                    return True  # Success means DB most likely already has the schema and the data loaded
+                except:
+                    return False  # Exception — we consider DB as empty (brand new, no schema, no data)
+
+
+def load_schema():  # Loading schema
+    this_script_dir = os.path.dirname(__file__)  # Look for a path to this script, load_data.py
+    with TypeDB.core_client(config.typedb_server_addr) as client:  # Establishing connection
+        with client.session(config.db, SessionType.SCHEMA) as session:  # Access data in the database
+            with open(os.path.join(this_script_dir, "../schema.tql"), "r") as schema:  # Read the schema.tql file
+                define_query = schema.read()
+                with session.transaction(TransactionType.WRITE) as transaction:  # Open transaction to write
+                    try:
+                        transaction.query().define(define_query)  # Execute query to load the schema
+                        transaction.commit()  # Commit transaction
+                        print("Loaded the " + config.db + " schema.")
+                        return True  # Setup complete
+                    except Exception as e:
+                        print("Failed to load schema: " + str(e))
+                        return False  # Setup failed
+
+
+def main():  # This is the main function of this script
+    with TypeDB.core_client(config.typedb_server_addr) as client:  # Establishing connection
+        if client.databases().contains(config.db):  # Check the DB existence
+            print("Detected DB " + config.db + ". Connecting.")
+            if not has_existing_data():  # Most likely the DB is empty and has no schema
+                print("Attempting to load the schema and data.")
+                if load_schema():  # Schema has been loaded
```

Comment:
OK, the calls to `core_client` look a lot cleaner now that we've introduced the `typedb_server_addr` config - now let's also simplify the implementation!

In the future, our Docs will contain detailed guidelines on when to create Clients, Sessions and Transactions and what their lifetimes should be. Until then, it's all our internal knowledge. And basically, a Client should, ideally, be created once in the whole lifetime of an _application_. That's what it's conceptually meant to represent (and it holds onto a single persistent connection with the server under-the-hood; this consumes resource, and we rarely need more than one.)

Currently, we open a client at the start of the `main` method, but then we open another client in `load_schema` (and another in `has_existing_data`, etc.)

Let's refactor all other methods in this file to take in a `client` parameter. Then, `main` will create one TypeDB client, and reuse it throughout the entire lifetime of the application.

I'd then do the same thing in `requests.py`. In this case, create the TypeDB client in the `main` method. Of course, there is the possibility that the user just exits the application and never _uses_ the client, but this would be an edge case scenario.

Reply:
Good idea. Fixed in https://github.com/vaticle/typedb-examples/commit/090229bc825d7a880fc38cb66428bc2bf2c1ee36